src/planner/binder/statement/bind_drop.cpp

```diff
@@ -43,10 +46,33 @@ BoundStatement Binder::Bind(DropStatement &stmt) {
 		stmt.info->schema = entry->schema->name;
 		break;
 	}
-	case CatalogType::DATABASE_ENTRY:
-		// attaching and detaching is read-only
-		stmt.info->catalog = SYSTEM_CATALOG;
+	case CatalogType::DATABASE_ENTRY: {
+		// use DuckDB default impl if no storage extension is registered to handle this functionality
+		auto &base = (DropInfo &)*stmt.info;
+		string database_name = base.name;
+
+		auto &config = DBConfig::GetConfig(context);
+
+		if (config.storage_extensions.empty()) {
+			// attaching and detaching is read-only
+			stmt.info->catalog = SYSTEM_CATALOG;
+		}
+		// for now assume only one storage extension provides the custom drop_database impl
+		for (auto &extension_entry : config.storage_extensions) {
```

Comment:
This is already used by DuckDB for `DETACH` - the storage extension does not seem to have any effect here. It creates a plan (sets `result.plan`) which then gets overriden by the regular `DETACH` afterwards.

Having the same code path handle two different cases (`detach` and `drop database`) also seems very error-prone, and might lead to problems where calling `detach` can lead to a database being dropped instead.

Reply:
Just to make sure I understand - IMO `detach` is something that's fully DuckDB owned (as it changes the databased registrations) and extension can be notified about it. `drop` (which implies detach) is something that storage extensions are *supposed* to implement as it requires custom deletion. @Mytherin @rjatwal am I thinking about this correctly?

Reply:
Yes. Mark and I discussed this being part of the changes in this PR